_posts/2019-03-19-til-ecto-validations-and-constraints.md

```diff
@@ -0,0 +1,184 @@
+
+### Constraints
+If we use validations, why would we also need constraints? Let's think about many apps that we use. When we sign up for an application, are we allowed to signup with the same email as another user? (Hint: the answer should always be no.)
+
+So why couldn't we just have a validation for uniqueness? Remember that, by definition validation are executed _prior to checking the database_. If we were to have a validation for uniqueness, that would mean everything is unique even if you're adding duplicates since it doesn't look at the database.
+
+A constraint is a rule that is enforced **by the database**. Your application will run through Ecto.Changeset's validations first without interacting with the database. Then it will execute any constraints by checking the database.
+
+Let's add our first constraint to enforce a user's unique email!
+```elixir
+user
+|> cast(attrs, [:display_name, :email, :password])
+|> validate_required([:display_name, :email, :password])
+|> unique_constraint(:email)
+```
+
+If we try to add the user to our database, all looks good the first time:
+```elixir
+iex(2)> user = %User{}
+iex(3)> attrs = %{display_name: "prince", email: "prince@test.com", password: "super_secret"}
+iex(4)> user |> User.changeset(attrs) |> Repo.insert()
+{:ok,
+ %MyCoolWebApp.Accounts.User{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
+   display_name: "prince",
+   email: "prince@test.com",
+   id: 1,
+   inserted_at: ~N[2019-03-18 01:41:34],
+   password: "super_secret",
+   password_hash: "$argon2i$v=19$m=65536,t=6,p=1$bhjgmBs9/gYcM2L5Z5sL/g$Z+4D7NIaauU+jwhdYRY4hz0adUdhjAJK6CwYk1AOJdE",
+   updated_at: ~N[2019-03-18 01:41:34]
+ }}
+```
+
+We want to make sure no duplicates get saved, so let's try sending the same thing again:
+```elixir
+iex(2)> user = %User{}
+iex(3)> attrs = %{display_name: "prince", email: "prince@test.com", password: "super_secret"}
+iex(4)> user |> User.changeset(attrs) |> Repo.insert()
+{:ok,
+ %MyCoolWebApp.Accounts.User{
+   __meta__: #Ecto.Schema.Metadata<:loaded, "users">,
+   display_name: "prince",
+   email: "prince@test.com",
+   id: 2,
+   inserted_at: ~N[2019-03-18 01:43:57],
+   password: "super_secret",
+   password_hash: "$argon2i$v=19$m=65536,t=6,p=1$H+Fq/IPW+M0YPHOZxMs13Q$ne+jDkwfcOigT8TKDIBYJjVwNdaNkzF/hc7YcRXRItY",
+   updated_at: ~N[2019-03-18 01:43:57]
+ }}
+```
+
+That's weird. It didn't show us an error? That's because Ecto doesn't know it is supposed to show an error. If you add a constraint to your `changeset/3`, you **must** enforce the constraint at a database level. So for our `unique_constraint`, we need to make sure to create a `unique_index` for the email field. Even though you wrote `unique_constraint` in the `changeset/3`, it doesn't check for that constraint unless there is a `unique_index` applied to the database.
```

Comment:
I would clarify that not only did it not "show" an error (which makes me think: error message added to changeset), but it didn't encounter or throw an error either. It successfully saved the record!